lib/muzic/src/zlib_comp.c

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright 2022 Young Mei
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include "uzlib.h"
+#include "zlib.h"
+
+void outbits(struct uzlib_comp* out, unsigned long bits, int nbits);
+
+static void out4bytes(struct uzlib_comp* out,
+                      unsigned char st,
+                      unsigned char nd,
+                      unsigned char rd,
+                      unsigned char th) {
+  int newlen = out->outlen + 4;
+  if (newlen > out->outsize) {
```

Comment:
This is going to be extremely costly. Is there no way to determine how many bytes might be needed and then make 1 large dynamic allocation?

Reply:
This bothered me as well. The underlying deflate implementation comes from uzlib, and it used a dynamically growing buffer for output. The size of this buffer grows linearly instead of exponentially (and what even the worse is it doesn't check if `realloc` succeed or not). I recognized the performance impact but decided to leave it untouched back then, to focus on the compatibility layer with Zlib. It worth some effort to rewrite the output buffer of uzlib's deflate implementation.

Reply:
I don't think this is something worth optimizing now either. Functionality first. I would make a note of this in #91 too. Also, please remember to make a note about the 32kB buffer size in #91. Anything that is essentially converted to a deferred problem or "for later optimization" should be explicitly noted (just so it is not forgotten).

Reply:
Sure. In fact, I have implemented the scalable buffer yesterday, and it doesn't require a 32kB buffer (from the beginning) now. But I think it's better to make it into another PR, as there are possibly some improvements for the current implementation. I'll make a note for this first.